## 1) Importing Data

In [2]:
import pandas as pd
import numpy as np

In [3]:
df=pd.read_csv('../data/Superstore_prepared.csv',encoding='Latin1')
df.head()

,Sales,Quantity,Discount,Profit,Delivery Time,Year of order,Month of order,Weekday of order,Ship Mode_Same Day,Ship Mode_Second Class,...,Sub-Category_Envelopes,Sub-Category_Fasteners,Sub-Category_Furnishings,Sub-Category_Labels,Sub-Category_Machines,Sub-Category_Paper,Sub-Category_Phones,Sub-Category_Storage,Sub-Category_Supplies,Sub-Category_Tables
0,261.9600,2,0.00,41.9136,3,2016,11,1,False,True,...,False,False,False,False,False,False,False,False,False,False
1,731.9400,3,0.00,219.5820,3,2016,11,1,False,True,...,False,False,False,False,False,False,False,False,False,False
2,14.6200,2,0.00,6.8714,4,2016,6,6,False,True,...,False,False,False,True,False,False,False,False,False,False
3,957.5775,5,0.45,-383.0310,7,2015,10,6,False,False,...,False,False,False,False,False,False,False,False,False,True
4,22.3680,2,0.20,2.5164,7,2015,10,6,False,False,...,False,False,False,False,False,False,False,True,False,False


## 2) import necessary ML modules

In [4]:
from sklearn.model_selection import train_test_split
# Models that will be trained
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import GradientBoostingRegressor

## 3) Split the dataset in training and testing

In [5]:
# Select the predictor set and target set
X=df.drop("Profit", axis=1)
Y=df["Profit"]

In [6]:
Xtrain,Xtest,Ytrain,Ytest=train_test_split(X,Y,test_size=0.2,random_state=42)

## 4) Scaling of predictor columns

In [7]:
from sklearn.preprocessing import StandardScaler

scaler=StandardScaler()
Xtrain_s=scaler.fit_transform(Xtrain)
XTest_s=scaler.transform(Xtest)

In [8]:
# we will use the scaled data for Linear Regression

## 5) Linear regression

In [9]:
lr=LinearRegression()
lr.fit(Xtrain_s,Ytrain)
ypred_lr=lr.predict(XTest_s)

## 6) Decision Tree Regression

In [10]:
dt=DecisionTreeRegressor(random_state=42)
dt.fit(Xtrain,Ytrain)
ypred_dt=dt.predict(Xtest)

## 7) Random Forest

In [11]:
rf=RandomForestRegressor(random_state=42)
rf.fit(Xtrain,Ytrain)
ypred_rf=rf.predict(Xtest)

## 8) Gradient Boosting

In [12]:
gb=GradientBoostingRegressor(random_state=42)
gb.fit(Xtrain,Ytrain)
ypred_gb=gb.predict(Xtest)

## 9) Linear Regression with PCA

In [13]:
from sklearn.decomposition import PCA

pca=PCA(n_components=0.9)

Xtrain_pca=pca.fit_transform(Xtrain_s)
Xtest_pca=pca.transform(XTest_s)

In [14]:
lr2=LinearRegression()
lr2.fit(Xtrain_pca,Ytrain)
ypred_lr_pca=lr2.predict(Xtest_pca)

## 10) Model Evaluation

In [35]:
from sklearn.metrics import mean_absolute_error,mean_squared_error,r2_score

#storing all results
results = []

In [36]:
# For Linear regression

mae = mean_absolute_error(Ytest, ypred_lr)
mse = mean_squared_error(Ytest, ypred_lr)
rmse = np.sqrt(mse)
r2 = r2_score(Ytest, ypred_lr)

In [ ]:
results.append({
    "Model": "Linear Regression",
    "MAE": mae,
    "RMSE": rmse,
    "R²": r2,
})

NameError: name 'acc' is not defined

In [18]:
# For Decision Tree

mae = mean_absolute_error(Ytest, ypred_dt)
mse = mean_squared_error(Ytest, ypred_dt)
rmse = np.sqrt(mse)
r2 = r2_score(Ytest, ypred_dt)

In [19]:
results.append({
    "Model": "Decision Tree Regressor",
    "MAE": mae,
    "RMSE": rmse,
    "R²": r2
})

In [20]:
# For Random Forest Regressor

mae = mean_absolute_error(Ytest, ypred_rf)
mse = mean_squared_error(Ytest, ypred_rf)
rmse = np.sqrt(mse)
r2 = r2_score(Ytest, ypred_rf)

In [21]:
results.append({
    "Model": "Random Forest Regressor",
    "MAE": mae,
    "RMSE": rmse,
    "R²": r2
})

In [22]:
# For Gradient Boosting regressor

mae = mean_absolute_error(Ytest, ypred_gb)
mse = mean_squared_error(Ytest, ypred_gb)
rmse = np.sqrt(mse)
r2 = r2_score(Ytest, ypred_gb)

In [23]:
results.append({
    "Model": "Gradient Boosting Regressor",
    "MAE": mae,
    "RMSE": rmse,
    "R²": r2
})

In [24]:
# For Linear regression with PCA

mae = mean_absolute_error(Ytest, ypred_lr_pca)
mse = mean_squared_error(Ytest, ypred_lr_pca)
rmse = np.sqrt(mse)
r2 = r2_score(Ytest, ypred_lr_pca)

In [25]:
results.append({
    "Model": "Linear Regression With PCA",
    "MAE": mae,
    "RMSE": rmse,
    "R²": r2
})

In [26]:
# Final Comparison

results_df=pd.DataFrame(results)
results_df

,Model,MAE,RMSE,R²
0,Linear Regression,71.010738,326.592784,-0.311676
1,Decision Tree Regressor,33.582199,291.727906,-0.046572
2,Random Forest Regressor,29.555235,266.521024,0.126473
3,Gradient Boosting Regressor,36.595450,252.558815,0.215599
4,Linear Regression With PCA,82.152892,291.048501,-0.041703


In [ ]:
# we prefer highest Rsquared value and lowest RMSE
#so Gradient Boosting Regressor is the clear choice for us

## Conclusion

- Five regression models were evaluated.
- Gradient Boosting Regressor achieved the best performance (R² = 0.216, RMSE = 252.56).
- Random Forest was the second-best model.
- Linear Regression struggled, indicating the relationship between features and profit is non-linear.
- PCA improved Linear Regression but did not outperform ensemble tree methods.